libkbfs/prefetcher.go

```diff
@@ -0,0 +1,168 @@
+package libkbfs
+
+import (
+	"io"
+	"sort"
+	"sync"
+
+	"golang.org/x/net/context"
+)
+
+const (
+	defaultIndirectPointerPrefetchCount int = 20
+	fileIndirectBlockPrefetchPriority   int = -100
+	dirEntryPrefetchPriority            int = -200
+)
+
+type prefetchRequest struct {
+	priority int
+	kmd      KeyMetadata
+	ptr      BlockPointer
+	block    Block
+}
+
+var _ Prefetcher = (*blockPrefetcher)(nil)
+
+type blockPrefetcher struct {
+	retriever  blockRetriever
+	progressCh chan prefetchRequest
+	shutdownCh chan struct{}
+	doneCh     chan struct{}
+	sg         sync.WaitGroup
+}
+
+func newBlockPrefetcher(retriever blockRetriever) *blockPrefetcher {
+	p := &blockPrefetcher{
+		retriever:  retriever,
+		progressCh: make(chan prefetchRequest),
+		shutdownCh: make(chan struct{}),
+		doneCh:     make(chan struct{}),
+	}
+	go p.run()
+	return p
+}
+
+func (p *blockPrefetcher) run() {
+runloop:
+	for {
+		select {
+		case req := <-p.progressCh:
+			if p.retriever == nil {
+				continue
+			}
+			ctx, cancel := context.WithCancel(context.Background())
+			ch := p.retriever.Request(ctx, req.priority, req.kmd, req.ptr, req.block, TransientEntry)
+			p.sg.Add(1)
+			go func() {
+				defer p.sg.Done()
+				select {
+				case _ = <-ch:
+				case <-p.shutdownCh:
+					// Cancel but still wait so p.doneCh accurately represents
+					// whether we still have requests pending.
+					cancel()
+					<-ch
+				}
+			}()
+		case <-p.shutdownCh:
+			break runloop
+		}
+	}
+	p.sg.Wait()
+	close(p.doneCh)
+}
+
+func (p *blockPrefetcher) request(priority int, kmd KeyMetadata, ptr BlockPointer, block Block) error {
+	select {
+	case p.progressCh <- prefetchRequest{priority, kmd, ptr, block}:
+		return nil
+	case <-p.shutdownCh:
+		return io.EOF
+	}
+}
+
+func (p *blockPrefetcher) prefetchIndirectFileBlock(b *FileBlock, kmd KeyMetadata, priority int) {
+	// Prefetch the first <n> indirect block pointers.
+	// TODO: do something smart with subsequent blocks.
+	numIPtrs := len(b.IPtrs)
+	if numIPtrs > defaultIndirectPointerPrefetchCount {
+		numIPtrs = defaultIndirectPointerPrefetchCount
+	}
+	for _, ptr := range b.IPtrs[:numIPtrs] {
+		p.request(fileIndirectBlockPrefetchPriority, kmd,
+			ptr.BlockPointer, b.NewEmpty())
+	}
+}
+
+func (p *blockPrefetcher) prefetchIndirectDirBlock(b *DirBlock, kmd KeyMetadata, priority int) {
+	// Prefetch the first <n> indirect block pointers.
+	numIPtrs := len(b.IPtrs)
+	if numIPtrs > defaultIndirectPointerPrefetchCount {
+		numIPtrs = defaultIndirectPointerPrefetchCount
+	}
+	for _, ptr := range b.IPtrs[:numIPtrs] {
```

Comment:
D'oh, I meant to put this comment below in prefetchIndirectDirBlock -- there is no such thing right now.